### ISSUE TITLE
Question: Why can't nameof be used with protected or private members?

### ISSUE THREAD

**Issue by jrshoare** [dotnet/csharplang#1990]
The `nameof` operator only works with class members that are public. This means that when used with reflection methods the member name needs to be specified as a string making it fragile for refactoring.

The following code snippet uses the unit testing `PrivateObject` class to hide the reflection code but demonstrates what I mean:

```csharp
void Main()
{
    Test t = new Test();
    Microsoft.VisualStudio.TestTools.UnitTesting.PrivateObject t2 =
        new Microsoft.VisualStudio.TestTools.UnitTesting.PrivateObject(t);

    Console.WriteLine(t.Method0());				// -> 0
    Console.WriteLine(t2.Invoke(nameof(Test.Method0)));
    Console.WriteLine(t2.Invoke("Method1"));			// -> 1
    //Console.WriteLine(t2.Invoke(nameof(Test.Method1)));	// CS0122, 'method' is inaccessible due to its protection level
    Console.WriteLine(t2.Invoke("Method2"));			// -> 2
    //Console.WriteLine(t2.Invoke(nameof(Test.Method2)));	// CS0122, 'method' is inaccessible due to its protection level
	
    Console.WriteLine(t.Property0);				// -> 0
    Console.WriteLine(t2.GetProperty(nameof(Test.Property0)));
    Console.WriteLine(t2.GetProperty("Property1"));		// -> 1
    //Console.WriteLine(t2.GetProperty(nameof(Test.Property1))); // CS0122, 'property' is inaccessible due to its protection level
    Console.WriteLine(t2.GetProperty("Property2"));		// -> 2
    //Console.WriteLine(t2.GetProperty(nameof(Test.Property2))); // CS0122, 'property' is inaccessible due to its protection level
    t.Property3 = 3;
    t2.SetProperty("Property3", 3);				// -> 3
    t2.SetProperty(nameof(Test.Property3), 3);			// -> 3
    Console.WriteLine(t2.GetProperty(nameof(Test.Property3)));	// -> 42
}

public class Test
{
    public int Method0() => 0;
    protected int Method1() => 1;
    private int Method2() => 2;
	
    public int Property0 => 0;
    protected int Property1 => 1;
    private int Property2 => 2;
	
    public int Property3 { private get => 42; set => Console.WriteLine(value); }
}
```

`Test.Property3` is an interesting case. Here the property `get` is private whilst the `set` inherits the `public` accessibility, yet I can use `nameof` on this member. I can see why as the 'outer' acccessibility modifier trumps the 'inner' but it still suggests confusing behaviour to me as the compiler does not complain about 'mixed' accessibility members.

Perhaps someone can suggest the rationale for limiting `nameof` to public members only. I would like to understand why this was decided as it only generates a compile-time string.

**Comment by gafter** [comment]
`nameof` can be used with any *accessible* member.

**Comment by svick** [comment]
If you used `namof` on a member and that member then changed, your code would no longer compile. That might make sense for reflection, but `nameof` is a general feature and in general, it's really bad if your code stops working because the implementation details of some other code changed.

**Comment by CyrusNajmabadi** [comment]
Because now it would be a breaking change if someone removed/renamed a private member.

**Comment by theunrepentantgeek** [comment]
> `nameof` can be used with any _accessible_ member.

This makes sense to me - it means that **nameof** can't be used to _increase_ your coupling to the target type.

**Comment by jnm2** [comment]
`nameof` can be useful for reflection, but it's not *about* reflection. It's about referring to the name used to address a C# language element. There is no name that can be used to address an inaccessible C# language element, so the attempt is inherently meaningless and the compile error is appropriate.

For example, `nameof(List<int>)` is rightfully `List`, the C# name, while `typeof(List<>).Name` is ``List`1``, the virtual machine representation which is what reflection is concerned with.

**Comment by willard720** [comment]
I mean, if you look at `nameof` as being a function that accepts a parameter, then it *wouldn't* make sense for it to use a non-public variable as no function can do that for obvious reasons, so why should `nameof` be different?

**Comment by SilentSin** [comment]
@svick

> If you used namof on a member and that member then changed, your code would no longer compile.

If you used `nameof` on a `public` member and that member then changed, your code would no longer compile. That is *exactly* what you want to happen when using `nameof` and it would still be what you want when using it on a non-public member.

> That might make sense for reflection, but nameof is a general feature and in general, it's really bad if your code stops working because the implementation details of some other code changed.

That's an argument against allowing/using reflection to target non-public members in the first place, not an argument against allowing compile time safety when doing so. Bottom line is: it can already be done, so why not allow it to be done safely?

@CyrusNajmabadi

> Because now it would be a breaking change if someone removed/renamed a private member.

I'd **much** rather a breaking change than a runtime error that might slip through the cracks. There are infinite possibilities for changes in a dependency to break the dependant code without changing the public API itself, and a compile error in this case would be far easier to identify and fix than any of them.

@theunrepentantgeek

> This makes sense to me - it means that nameof can't be used to increase your coupling to the target type.

Allowing `nameof` on non-publics wouldn't increase that coupling either. You are already depending on a member in the target type when using reflection on it (or whatever it is you want to do with the name). The only difference is whether you get a compile error or a runtime error.

Forcing people to use magic strings seems more like the Javascript philosophy of letting pretty much anything run and giving unintended results than the C# philosophy of giving as much compile time safety as possible.

**Comment by CyrusNajmabadi** [comment]
> If you used nameof on a public member and that member then changed, your code would no longer compile. 

Yes.  It's known that changing a public member is a breaking change.  That's why it isn't done.  But this would make it so that changing private members would now be breaking changes as well.  

> I'd much rather a breaking change

You would.  But many teams who can't ship breaking changes would now be screwed.  

--

Also, none of this works in any sort of world where reference-assemblies are used.

**Comment by SilentSin** [comment]
> You would. But many teams who can't ship breaking changes would now be screwed.

If you make a plugin that I depend on and I'm using reflection against its private members and you aren't allowed to break my code, then you already can't change those members. This wouldn't change that fact.

> Also, none of this works in any sort of world where reference-assemblies are used.

Even if it only worked for members within the same assembly (as if they were internal) it would still be very useful.

**Comment by CyrusNajmabadi** [comment]
> Even if it only worked for members within the same assembly (as if they were internal) it would still be very useful.

If you're in the same assembly, the internal members would be accessible to you and there would be a way to reference accessible members in a supported manner.

**Comment by CyrusNajmabadi** [comment]
> If you make a plugin that I depend on and I'm using reflection against its private members and you aren't allowed to break my code, then you already can't change those members. 

Sure you can.  As the language exists today, those private members are an internal implementation detail.  You're asking to change teh language to make it no longer an internal impl detail.  It's now a spec'ed part of the language.  It would now be breaking, for example, to not include your internals in a reference assembly.

**Comment by SilentSin** [comment]
> If you're in the same assembly, the internal members would be accessible to you and there would be a way to reference accessible members in a supported manner.

Yes, but then you're making them accessible all over your code base when all you need is the name.

For example, in Unity the recommended way to write a custom inspector for a class is to keep your fields private and have the custom inspector class access their SerializedProperty by name. Making those fields internal and letting everything fiddle with them freely would be far worse than just using a magic string for the name. And not needing to use a magic string would be even better.

**Comment by yaakov-h** [comment]
If it's your own class, then you can just create a public static/const field that contains the name of the private member...

**Comment by CyrusNajmabadi** [comment]
If you want to just expose the name... then just expose the name.  It's your code :)

**Comment by SilentSin** [comment]
You can, but that's more work and still not as reliable as using `nameof`.

**Comment by jnm2** [comment]
Right now there are two kinds of member. There is the kind that is part of a contract (inputs and outputs of the system) and known at compile time, and there is the kind that is not part of a contract and should be expected to change freely at runtime as a black box without detriment to the contract.

By asking for both kinds of member to be pinned down from the outside by `nameof`, you're effectively asking for all implementation details to become public contract. Instead of two kinds of thing (API surface and implementation details) you're asking for there to be just one kind of thing.

I would say that both kinds of thing are useful and have their place and we shouldn't be trying to get rid of one. You can already make things public or internal. Even better, you can do API design so that everything you need is exposed and things don't become public contract in an accidental and haphazard fashion.

**Comment by SilentSin** [comment]
> Why not? how is it not reliable?
> You expose the name (using nameof internally if you want). Now it's clear what you're doing since you're explicitly stating that this is part of your public surface area. Now, if you change anything, things still work.

You're right, I wasn't thinking of using nameof internally. It's still more work though.

I've already got:
```
/// <summary>
/// Best int ever.
/// </summary>
[SerializeField]
private int _MyInt;
public int MyInt { get { return _MyInt; } }
```
And that would add yet another line to every field.

That doesn't sound too bad and I might try it next time I have the opportunity, but in practice I've *never* actually seen a custom inspector that uses constants for the field names like that in almost a decade of using Unity, [including in the source code of Unity itself](https://github.com/Unity-Technologies/UnityCsReference/blob/73f36bfe71b68d241a6802e0396dc6d6822cb520/Editor/Mono/Inspector/RigidbodyEditor.cs) (see the OnEnable method). While your suggestion has definite merit, the fact that no one actually does it tells me there's room for improvement.

**Comment by SilentSin** [comment]
> I want to expose the name. Not the ability to get ... it. I can't parse this. You want to expose the name... but not the name... ?

I want to expose the name, not the value.

**Comment by CyrusNajmabadi** [comment]
> I want to expose the name, not the value.

Right.  I got that.  So expose it with: `public static readonly string TheName = ...`

**Comment by jrshoare** [comment]
Thanks for all your comments. I now have a better picture of why the decision is what it is and further, what I can do to mitigate the lack of accessibility should I need to get at the inaccessible parts for testing.

**Comment by jon49** [comment]
Another use for exposing protected/private members to `nameof` would be for testing and mocking. I'm mocking an item and I don't want the part that I'm testing to go down the rabbit hole of inheritance and so `Moq` requires the string name of the inherited method. Ideally I would be able to use `nameof(myProtectedMethod)` to do this. But it seems strange to force the API of the abstract class to expose the name of the method just so I can test it without worry about the name changing in the future.

But whatever. It sounds like it will be kept for only public members.

**Comment by jnm2** [comment]
> But it seems strange to force the API of the abstract class to expose the name of the method just so I can test it without worry about the name changing in the future.

The alternatives are even stranger...

**Comment by jon49** [comment]
@jnm2 , Ha ha, yes, I guess the whole concept of mixing side affects and logic is a bit strange which then requires strange testing with lots of mocks :-).

**Comment by voroninp** [comment]
@svick ,
>it's really bad if your code stops working because the implementation details of some other code changed.

It depends a lot! 

When I define relationships in EF Core, and navigation property is private, I indeed want my code to break at compile time rather than at run time.  I'd prefer to have `WithMany(nameof(SomeClass.SomePrivateProperty))` instead of `WithMany("SomePrivateProperty")`. This is the case where I need coupling.

Moreover, your argument is applicable to any accessible class member. If it is changed, code usually does not compile. I find it as a questionable design decision not to let `nameof` see private members.

I'd be ok with an alternative like `nameof?(SomeClass.SomePrivateMember)` or `nameof!(SomeClass.SomePrivateMember)`

**Comment by CyrusNajmabadi** [comment]
> I'd be ok with an alternative

The alternative is to just expose the name of the member.  i.e.  `public const string SomePrivateMemberName = nameof(SomePrivateMember);`.

**Comment by voroninp** [comment]
@CyrusNajmabady, not that good alternative. Because as soon as you rename your property you’ll get a mismatch between `SomePrivateMemberRenamed` and `SomePrivateMemberName`.

**Comment by CyrusNajmabadi** [comment]
right. but none of your consumptive code will break. that's the benefit.  `SomePrivateMemberName` will still have the right value since it uses 'nameof' in its initializer.

**Comment by douglasg14b** [comment]
All this talk about extra coupling, but in the case of tests that coupling is wanted...

Have a class with a method that requires a private field be set, else that fails? (Think EF, DDD, and collections). It would sure be nice to have a type safe way to refer to that field's name instead of as a magic string....

**Comment by theunrepentantgeek** [comment]
@douglasg14b wrote

> All this talk about extra coupling, but in the case of tests that coupling is wanted...

There are those (including myself) who prefer to write tests based on the _external_ behaviour of the type in question; if there are too many internal details that I find I want to test, that motivates me to find the missing abstraction to pull out as a new type.

Even if we grant the desirability of writing unit tests against implementation details, adding a language feature to support that niche scenario seems unwise, especially given the myriad of ways that `nameof(_member)` could cause problems.

**Comment by voroninp** [comment]
@theunrepentantgeek I'd say that visibility was either broken from the very beginning, or it is always not enough to cover all use cases. The existence of `InternalsVisibleToAttribute` is a proof.  `private protected` modifier was added to C# only in version 7.2 which also indicates that people needed more options.

Is this feature that dangerous and expensive? Call this operator `nameof!` to indicate that it is different from the less powerful `nameof`.

**Comment by chrisoverzero** [comment]
> Is this feature that dangerous […]?

Yes, it makes refactoring nightmare-shaped, knowing that consumers could be relying on the names of things in your internals. The `InternalsVisibleTo` attribute is opt-in for the library author. Various access modifiers are opt-in for the library author. None of these are under the control of the library consumer. 

> Honestly, I'd prefer having a `God mode` when all object intrinsics are accessible to my code.

That’s the `public` keyword. If you want to spill your guts all over your callers, there’s nothing in C# stopping you. If you want others to spill their guts all over you, you require their opt-in.

**Comment by voroninp** [comment]
> Yes, it makes refactoring while maintaining backwards compatibility nightmare-shaped, knowing that consumers could be relying on the names of things in your internals.

Well, can't you say the same about reflection? You as an author cannot prevent anyone to use reflection and compiled expression for creating setters even for private readonly properties.
But should you really care?

> The `InternalsVisibleTo` attribute is opt-in for the library author and targeted to particular assemblies. Various access modifiers are also opt-in for the library author. 

Ok, then where is my `PrivatesVisibleTo`? ;-)

```CSharp
[PrivatesVisibleTo("AssemblyQualifiedClassName" or Type)]
public class Foo
{
    private readonly int _filed1 = 1;
}
```
There would be an opt-in option.
And it is useful for having nested public classes with private constructors accessible to the nesting class.

**Comment by HaloFour** [comment]
> `private protected` modifier was added to C# only in version 7.2 which also indicates that people needed more options.

That was originally going to be added in C# 6.0, but there was so much arguing over it, a lot of which was because it was so pointless.  But keyword choice aside it was an incredibly easy add because the CLR has supported that accessibility modifier since 1.0, so no runtime changes were necessary.

> Well, can't you say the same about reflection?

Reflection is inherently brittle.  The language doesn't offer any additional support to make it easier to use, and that's on purpose.  You rely on reflection at your own risk.

> Ok, then where is my `PrivatesVisibleTo`? ;-)

There's no need for it, you can already mark the constructor `internal` and make use of `InternalsVisibleTo` for testing assemblies.

**Comment by voroninp** [comment]
> There's no need for it, you can already mark the constructor internal and make use of `InternalsVisibleTo` for testing assemblies.

Well, it is a bit weird that I need to compromise on visibility and make something more accessible within assembly to make it accessible in other assembly.
You should also look at it from the point of view of the developer working in a team.
The intent matters a lot. I need to clearly set the boundaries, and if something is private then it is private for everything besides some chosen pieces of the codebase. Assembly is too large unit for that. =(

**Comment by CyrusNajmabadi** [comment]
> Is this feature that dangerous and expensive

No. It simply doesn't work. For one, tons of assembles ship with no private details, and that is becoming more and more the norm. 

So it's not something that can even be effectively done.

**Comment by CyrusNajmabadi** [comment]
> Well, can't you say the same about reflection?

Nameof is not reflection. One is part of the c# language, one exists entirely outside of it.

**Comment by juliusfriedman** [comment]
You can also mark the member as `internal` or `protected internal`

**Comment by HaloFour** [comment]
@voroninp 

> Assembly is too large unit for that.

That sounds like part of the problem.  Break apart the assemblies into smaller pieces and that becomes less of a concern.  Or write an analyzer which limits calling those `internal` members from non-sanctioned code within the same assembly.  C# provides plenty of accessibility options, adding more so that you can limit access except where you feel like having "God-mode" enabled makes little sense.

**Comment by kaleidocore** [comment]
All the suggestions and variations on adding some static field to expose the nameof(member) are unbelievably stupid.

1. If a class designer expects the consumer to be interested in some particular member he should obviously **not** make it private.

2. The only reason to explicitly expose the _name_ of a member is if you expect the consumer to _reflect_ that explicit member, in which case, again, there is **no** reason for the member to be private in the first place!

3. The need to reflect hidden members typically arises from **_not_** having access to the source you are reflecting.


I will bet you my house, my dog and my firstborn child that nameof() is almost exclusively used for:
1. Debug information, including logging and technical error messages
2. Reflection

In both of these cases private members would be fine and useful to expose to nameof() because the alternative - _magical strings_  - is just so much worse. The whole reason and intent of using nameof() rather than magical strings is _precisely_ so that things break at _compile time_ rather than runtime. Reflection is an incredibly powerful tool and people **will** and **do** reflect hidden members, whether the original designer intended it or not.

If you absolutely need to hide "unpublished" API details from the simpleton consumers you could still do that by (for example) requiring a fully qualified identifier to access hidden members, i.e:
`nameof(MyClass.MyPublicProp)`
vs
`nameof(global::MyNamespace.MyClass.MyPrivateProp)`

The bottom line is: I have seen tons and tons of ugly runtime failures because someone reflected something by magical strings and either misspelled it or forgot to update the magical string after refactoring or swapping assemblies. All of that could be easily and harmlessly fixed if just nameof() allowed it.

**Comment by voroninp** [comment]
@kaleidocore In the light of assembly trimming `nameof` becomes even more important, because compiler may have an chance to inject additional metadata to avoid dropping the part which seem to be not used otherwise ;-)

**Comment by jnm2** [comment]
> All the suggestions and variations on adding some static field to expose the nameof(member) are unbelievably stupid.

This violates the https://github.com/dotnet/csharplang/blob/master/CODE-OF-CONDUCT.md.

**Comment by jnm2** [comment]
@kaleidocore 

> I will bet you my house, my dog and my firstborn child that nameof() is almost exclusively used for:
> 
>     1. Debug information, including logging and technical error messages
>     2. Reflection

One of the most major uses I've seen is when switching on PropertyChangedEventArgs.PropertyName. Another pretty common use is defining string constants.

**Comment by theunrepentantgeek** [comment]
@kaleidocore it seems you are focusing entirely on the question of _"Would this feature solve the stated problem?"_ while minimizing (or completely ignoring) _"What problems would occur as a consequence of this feature?"_. The C# Language Design team don't have that luxury - they have to consider _**all**_ the pros and cons of a feature.

>  I have seen tons and tons of ugly runtime failures because someone reflected something by magical strings and either misspelled it or forgot to update the magical string after refactoring 

This implies failure of testing far more than a limitation of the language. See my example below.

> or swapping assemblies. 

The **nameof** operator is turned into a literal string by the compiler, so you'd still get exactly the same failure by swapping assemblies.

> All the suggestions and variations on adding some static field to expose the nameof(member) are completely missing the point.

I don't think they are. 

Here's how I'd solve the problem. Let's assume there's a repository class that I don't control called `PersonRepository` and I need to get at the protected property `ConnectionString`.

Applying _separation of concerns_, I would define a static class to take care of all the reflection stuff, keeping it out of the core business code (this is the step you said was stupid):

``` csharp
public static class RepositoryMember
{
    private const string ConnectionStringPropertyName = "ConnectionString";

    public static string ConnectionString(PersonRepository person) { ... }
}
```

Next, I'd write unit tests to ensure this helper works as expected:

``` csharp
public class RepositoryMemberTests
{
    [Fact]
    public void ConnectionString_GivenInstance_ReturnsNonEmptyResult() { ... }
}
```

This ensures that the reflection works, and gives a local test failure (or, at worst, a test failure during continuous integration), avoiding production runtime issues. Even better, such a test failure clearly signals that the problem is with reflection, not elsewhere, helping to reduce the time to remedy.

Consuming the class is the last step:

``` csharp
var connectionString = RepositoryMember.ConnectionString(repository);
```

One final point ... this isn't a case of replacing a single `nameof()` with multiple additional classes. The separation of concerns and the testing is (in my opinion) still absolutely required. In my experience there's a lot more than just getting the magic string wrong that can go awry with reflection code, which is why it's necessary to isolate away from the mainline of code and apply specific tests.

**Comment by voroninp** [comment]
> What problems would occur as a consequence of this feature?

Indeed, I do like to know what the problems are, if syntax is extended to clearly indicate that I want to break the privacy of the class and get the name of its hidden member? Right now it feels more like being attached to dogmas rather than discussion of the possible issues for the C# dev team.

Btw, your example with encoding name of the member just as a string won't work, if code is obfuscated. And this means concerns are not that separated. ;-)

**Comment by theunrepentantgeek** [comment]
> Indeed, I do like to know what the problems are

I just skimmed the entire thread above and the problems are well articulated already. I don't think I'd be adding value by repeating them again.

**Comment by HaloFour** [comment]
@voroninp 

> Btw, your example with encoding name of the member just as a string won't work, if code is obfuscated.

If the assembly is obfuscated the `nameof` would be just as impacted.  The compiler captures the name of the referenced member at compile-time.  If you then obfuscate the assembly those names will still end up scrambled but the captured constants will still reference the original name.

**Comment by CyrusNajmabadi** [comment]
> Right now it feels more like being attached to dogmas rather than discussion of the possible issues for the C# dev team.

I'm a member of the C# ldm.  You can read my thoughts in the thread on not only why i think it's a bad idea, but hte literal technical limitations that make this impossible.  i.e. we have moved very heavily (and will continue to do so) in the direction that inaccessible members are not even exposed from teh dlls the compiler references.  So ignoring that i don't think it's a good idea on its face, it wouldn't even be something the compiler coudl support as it would have no idea if `nameof(X.Y)` could even bind to one of those members since the compiler would have no member to see.

**Comment by voroninp** [comment]
@CyrusNajmabadi 
>  in the direction that inaccessible members are not even exposed from teh dlls the compiler references.
Does it mean reflection won't work either?

**Comment by CyrusNajmabadi** [comment]
> Does it mean reflection won't work either?

No.  Reflection operates over the actual information available at runtime.  However, what is available at runtime is unrelated to what is available at compile time.  They are disparate sets of information.

**Comment by kaleidocore** [comment]
@theunrepentantgeek 
>The nameof operator is turned into a literal string by the compiler, so you'd still get exactly the same failure by swapping assemblies.

I'm talking about upgrading or downgrading referenced assemblies, compile-time - so nameof() would be properly updated.

>This implies failure of testing far more than a limitation of the language.
Unit testing adds complexity and overhead. It requires strict discipline among all participants, extra training, well-known and well-defined policies for creating unit tests, etc. All of that becomes even harder in a big, long-lived project with tons of developers, contractors and people coming and going. Sure, both manual and automated "testing" can solve many problems, but nothing beats having the compiler do the work for you.

>this isn't a case of replacing a single nameof() with multiple additional classes. The separation of concerns and the testing is (in my opinion) still absolutely required.
I see what you mean, but reality often disagrees.


@CyrusNajmabadi 
>we have moved very heavily (and will continue to do so) in the direction that inaccessible members are not even exposed from teh dlls the compiler references. So ignoring that i don't think it's a good idea on its face, it wouldn't even be something the compiler coudl support as it would have no idea if `nameof(X.Y)` could even bind to one of those members since the compiler would have no member to see.

Fair enough, that's a technical limitation which is obviously hard to argue against. To me, nameof() always felt like something in the twilight zone of preprocessing and compiling anyways. But then if I would try to access some hidden member by `X.Y`, how can the compiler tell me the member is inaccessible due to visibility, rather than claiming it's an unknown identifier?

**Comment by CyrusNajmabadi** [comment]
The compiler can still crack things open to give a better error message. That's different than deciding what the language would allow and how that would impact the entire .net tooling story :-)

**Comment by mikkelens** [comment]
Sad to see this was entirely forgotten/rejected because I still can't find a solution to the core problem of replacing magic strings with something better (except nameof() solves it, just not for private/inaccessible fields).

**Comment by voroninp** [comment]
@mikkelens , well, now it can be solved with source generators. Not perfect, but ok.

**Comment by mikkelens** [comment]
@voroninp I've heard of source generators, but I'm not aware how it solves this issue (maybe I don't understand it correctly). Can you link me to a reference that showcases a solution to a problem like the one I describe?

**Comment by voroninp** [comment]
@mikkelens , I'd think of generator for partial class which generates something like this:
```
partial class MyCass
{
    public static class Fields
    {
        public static read-only string _someField = "_someField";
    }
}
```
The only issue is that it's not refactoring friendly. As soon as you rename the field, this code will obviously become invalid:

`MyClass.Fields._someField`

But at least you have compiler safety.
